### ISSUE TITLE
Undocumented 'type' key in connection.response['entries'][0]

### ISSUE THREAD

**Issue by kpinc** [cannatag/ldap3#788]
After calling:

```
server = ldap3.Server.from_definition(
        server_name,
        'examples/dummy_server_info.json',
        'examples/dummy_server_schema.json')
connection = ldap3.Connection(
        server, user=user, password=password,
        return_empty_attributes=False,
	client_strategy=ldap3.MOCK_SYNC)
```
and then adding an entry, a search operation

```
connection.search(
	dn, '(objectclass=*)', attributes=ldap3.ALL_ATTRIBUTES)
```
 returns a "type" key/value in:

```
connection.response['entries'][0]
```

"type" and its value are undocumented.  This makes it more difficult to compare
the response entries with expected responses.

**Comment by cannatag** [comment]
type can be any of 'searchResEntry', 'searchResRef', 'intermediateResponse' as specified in RFC 4511. I agre that this is not documented but is in the LDAP standard.

**Comment by kpinc** [comment]
On Fri, 21 Feb 2020 09:33:22 -0800
Giovanni Cannata <notifications@github.com> wrote:

> type can be any of 'searchResEntry', 'searchResRef',
> 'intermediateResponse' as specified in RFC 4511. I agre that this is
> not documented but is in the LDAP standard.

The problem is not that the value is not documented, the problem
is that the key is not documented as something that could exist.
So when writing tests and looking at the expected response
the code sees an extra entry in the dict that shouldn't
exist.

Regards,

Karl <kop@karlpinc.com>
Free Software:  "You don't pay back, you pay forward."
                 -- Robert A. Heinlein

**Comment by cannatag** [comment]
You're right. I forgot to include this in the SEARCH operation docs (https://ldap3.readthedocs.io/searches.html). I will add the following:

```
* type: the type of the response as specified by RFC4511. It can be any of the following:
    - searchResEntry: the response contain an entry that satisfies the search filter
    - searchResRef: the response is a continuation referral to another DIT where the search should continue (usually handled automatically by ldap3)
    - searchResDone: search completed, no other entries found (handled automatically by ldap3)
    - intermediateResponse: a response in a persistent search (usually a continued search that never ends)
    - extendResp: a response to an extended request (some are automatically handled by ldap3)
```
Thanks.

ps. in the current MOCK strategies it should always contains ``searchResEntry``

**Comment by kpinc** [comment]
On Sat, 22 Feb 2020 09:02:47 -0800
Giovanni Cannata <notifications@github.com> wrote:

> You're right. I forgot to include this in the SEARCH operation docs
> (https://ldap3.readthedocs.io/searches.html). I will add the
> following:

Thanks.

FYI.  It would be really nice if the connection.response_to_json()
and entry.entry_to_json() returned "identical" results.  At least
so far as is possible, or could be made possible.  And whatever
other json related output operations.  I don't recall the details
but do recall bumping into some differences or another -- of
which the "type" key was only one.  (It may or may not have had
to do with "raw" as well.  I seem to recall different keys
used for the raw dict entries, somewhere.)

Regards,

Karl <kop@karlpinc.com>
Free Software:  "You don't pay back, you pay forward."
                 -- Robert A. Heinlein

**Comment by cannatag** [comment]
Response and Entry are different objects: the former is the response from the server, so the type and the raw_attributes are in the returned dict; the latter is from the Abstraction Layer, an high level representation of the data returned by the server. You can get the raw attribute with the entry_raw_attributes() method of the entry. There is no way to get the type of the entry because the entry represent the data (that can be heavly modified by formatters or de-references of other entries).

**Comment by kpinc** [comment]
On Sat, 22 Feb 2020 18:09:55 -0800
Giovanni Cannata <notifications@github.com> wrote:

> Response and Entry are different objects: the former is the response
> from the server, so the type and the raw_attributes are in the
> returned dict; the latter is from the Abstraction Layer, an high
> level representation of the data returned by the server.

Understood.

> You can get
> the raw attribute with the entry_raw_attributes() method of the
> entry.

You can (IIRC) also pass an (undocumented?) "raw" keyword argument
to the json dumper.

> There is no way to get the type of the entry because the entry
> represent the data (that can be heavly modified by formatters or
> de-references of other entries).

Yes.  "type" would be a case of where it's not feasible to have
identical json.  "type" is a property of a Response only.

Never the less, the entry portions of the json dumps of Response and
Entry objects represent the same data.  My point is that
when a particular data structure and identical data are represented
as json it really shouldn't matter what series of operations were used
to produce the json.  It is reasonable, and useful, to expect all
resulting json to always be "the same"; to compare equal -- at least in
so far as the dict produced when the json is parsed.  

(Obviously,
the printed json representations may not be identical for reasons
related to formatting, ordering, etc.)

Regards,


Karl <kop@karlpinc.com>
Free Software:  "You don't pay back, you pay forward."
                 -- Robert A. Heinlein

**Comment by cannatag** [comment]
Fixed in ldap3 v2.7.
